File: backend/src/main/java/com/example/backend/service/AuthService.java
package com.example.backend.service;

import com.example.backend.model.User;
import com.example.backend.model.UserRole;
import com.example.backend.repo.UserRepository;
import com.example.backend.util.JwtUtil;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;

@Service
public class AuthService {
    private final UserRepository userRepository;
    
    @Autowired
    private JwtUtil jwtUtil;
    
    public AuthService(UserRepository userRepository) {
        this.userRepository = userRepository;
        initializeDefaultUsers();
    }
    
    private void initializeDefaultUsers() {
        // Create default users if they don't exist
        System.out.println("DEBUG: Initializing default users...");
        if (userRepository.findByUsername("superadmin").isEmpty()) {
            System.out.println("DEBUG: Creating superadmin user...");
            User superadmin = new User();
            superadmin.setUsername("superadmin");
            superadmin.setPassword("admin123");
            superadmin.setFullName("Super Admin");
            superadmin.setRole(UserRole.SUPERADMIN);
            superadmin.setDepartment("IT");
            superadmin.setIsActive(true);
            User savedUser = userRepository.save(superadmin);
            System.out.println("DEBUG: Superadmin user created with ID: " + savedUser.getId());
        } else {
            System.out.println("DEBUG: Superadmin user already exists");
            // Update existing superadmin user to ensure correct password and status
            User existingUser = userRepository.findByUsername("superadmin").get();
            System.out.println("DEBUG: Existing superadmin - Password: " + existingUser.getPassword() + 
                             ", IsActive: " + existingUser.getIsActive());
            
            // Reset password and ensure user is active
            existingUser.setPassword("admin123");
            existingUser.setIsActive(true);
            userRepository.save(existingUser);
            System.out.println("DEBUG: Superadmin user updated with correct password and status");
        }
        
        if (userRepository.findByUsername("shambhu").isEmpty()) {
            User shambhu = new User();
            shambhu.setUsername("shambhu");
            shambhu.setPassword("shambhu123");
            shambhu.setFullName("Shambhu Sir");
            shambhu.setRole(UserRole.IT_MANAGER);
            shambhu.setDepartment("IT");
            userRepository.save(shambhu);
        }
        
        if (userRepository.findByUsername("joshi").isEmpty()) {
            User joshi = new User();
            joshi.setUsername("joshi");
            joshi.setPassword("joshi123");
            joshi.setFullName("Joshi Sir");
            joshi.setRole(UserRole.FINANCE_MANAGER);
            joshi.setDepartment("Finance");
            userRepository.save(joshi);
        }
        
        if (userRepository.findByUsername("employee1").isEmpty()) {
            User employee1 = new User();
            employee1.setUsername("employee1");
            employee1.setPassword("emp123");
            employee1.setFullName("John Doe");
            employee1.setRole(UserRole.EMPLOYEE);
            employee1.setDepartment("IT");
            userRepository.save(employee1);
        }
        
        if (userRepository.findByUsername("employee2").isEmpty()) {
            User employee2 = new User();
            employee2.setUsername("employee2");
            employee2.setPassword("emp123");
            employee2.setFullName("Jane Smith");
            employee2.setRole(UserRole.EMPLOYEE);
            employee2.setDepartment("Sales");
            userRepository.save(employee2);
        }
        
        // Create Department Managers
        if (userRepository.findByUsername("salesmanager").isEmpty()) {
            User salesManager = new User();
            salesManager.setUsername("salesmanager");
            salesManager.setPassword("sales123");
            salesManager.setFullName("Sales Manager");
            salesManager.setRole(UserRole.DEPARTMENT_MANAGER);
            salesManager.setDepartment("Sales");
            userRepository.save(salesManager);
        }
        
        if (userRepository.findByUsername("itmanager").isEmpty()) {
            User itManager = new User();
            itManager.setUsername("itmanager");
            itManager.setPassword("it123");
            itManager.setFullName("IT Department Manager");
            itManager.setRole(UserRole.DEPARTMENT_MANAGER);
            itManager.setDepartment("IT");
            userRepository.save(itManager);
        }
    }
    
    public Map<String, Object> login(String username, String password) {
        System.out.println("DEBUG: Login attempt for username: " + username);
        Optional<User> userOpt = userRepository.findByUsername(username);
        
        if (userOpt.isPresent()) {
            User user = userOpt.get();
            System.out.println("DEBUG: User found - Username: " + user.getUsername() + 
                             ", Password matches: " + user.getPassword().equals(password) + 
                             ", IsActive: " + user.getIsActive());
            
            if (user.getPassword().equals(password) && user.getIsActive()) {
                String token = jwtUtil.generateToken(username, user.getRole().name(), user.getDepartment(), user.getId());
                
                System.out.println("DEBUG: Login successful for user: " + username + ", JWT token generated");
                
                Map<String, Object> response = new HashMap<>();
                response.put("success", true);
                response.put("token", token);
                response.put("user", Map.of(
                    "id", user.getId(),
                    "username", user.getUsername(),
                    "fullName", user.getFullName(),
                    "role", user.getRole().name(),
                    "department", user.getDepartment()
                ));
                return response;
            } else {
                System.out.println("DEBUG: Login failed - Password mismatch or user inactive");
            }
        } else {
            System.out.println("DEBUG: User not found in database");
        }
        
        System.out.println("DEBUG: Login failed for username: " + username);
        return Map.of("success", false, "message", "Invalid credentials");
    }
    
    public User getCurrentUser(String token) {
        try {
            String username = jwtUtil.extractUsername(token);
            Optional<User> userOpt = userRepository.findByUsername(username);
            if (userOpt.isPresent() && userOpt.get().getIsActive()) {
                return userOpt.get();
            }
        } catch (Exception e) {
            System.out.println("DEBUG: Error extracting user from token: " + e.getMessage());
        }
        return null;
    }
    
    public void logout(String token) {
        // JWT tokens are stateless, so no server-side logout needed
        // Token will expire naturally
        System.out.println("DEBUG: User logged out (JWT token will expire naturally)");
    }
    
    public boolean hasPermission(String token, String requiredRole) {
        try {
            String role = jwtUtil.extractRole(token);
            if (role == null) return false;
            
            // SUPERADMIN has all permissions
            if (role.equals("SUPERADMIN")) return true;
            
            // Check specific role permissions
            return role.equals(requiredRole);
        } catch (Exception e) {
            System.out.println("DEBUG: Error checking permissions: " + e.getMessage());
            return false;
        }
    }
    
    public boolean isEmployee(User user) {
        return user != null && user.getRole() == UserRole.EMPLOYEE;
    }
    
    public boolean hasRole(User user, String roleName) {
        return user != null && user.getRole().name().equals(roleName);
    }
    
    public boolean isDepartmentManager(User user) {
        return user != null && user.getRole() == UserRole.DEPARTMENT_MANAGER;
    }
    
    public boolean canApproveDepartmentRequests(User user, String requestDepartment) {
        if (user == null) return false;
        
        // SUPERADMIN can approve any department
        if (user.getRole() == UserRole.SUPERADMIN) return true;
        
        // Department Manager can only approve their own department
        if (user.getRole() == UserRole.DEPARTMENT_MANAGER) {
            return user.getDepartment().equals(requestDepartment);
        }
        
        return false;
    }
    
    
    public Map<String, Object> createUser(String token, String username, String password, String fullName, String role, String department) {
        try {
            // Check if current user is SUPERADMIN
            User currentUser = getCurrentUser(token);
            if (currentUser == null || currentUser.getRole() != UserRole.SUPERADMIN) {
                return Map.of("success", false, "message", "Access denied. Only SUPERADMIN can create users.");
            }
            
            // Check if username already exists
            if (userRepository.findByUsername(username).isPresent()) {
                return Map.of("success", false, "message", "Username already exists.");
            }
            
            // Validate role
            UserRole userRole;
            try {
                userRole = UserRole.valueOf(role.toUpperCase());
            } catch (IllegalArgumentException e) {
                return Map.of("success", false, "message", "Invalid role. Valid roles: SUPERADMIN, IT_MANAGER, FINANCE_MANAGER, DEPARTMENT_MANAGER, EMPLOYEE");
            }
            
            // Create new user
            User newUser = new User();
            newUser.setUsername(username);
            newUser.setPassword(password);
            newUser.setFullName(fullName);
            newUser.setRole(userRole);
            newUser.setDepartment(department);
            newUser.setIsActive(true);
            
            User savedUser = userRepository.save(newUser);
            
            System.out.println("DEBUG: User created successfully: " + username + " with role: " + role);
            
            return Map.of(
                "success", true, 
                "message", "User created successfully",
                "user", Map.of(
                    "id", savedUser.getId(),
                    "username", savedUser.getUsername(),
                    "fullName", savedUser.getFullName(),
                    "role", savedUser.getRole().name(),
                    "department", savedUser.getDepartment(),
                    "isActive", savedUser.getIsActive()
                )
            );
            
        } catch (Exception e) {
            System.out.println("DEBUG: Error creating user: " + e.getMessage());
            e.printStackTrace();
            return Map.of("success", false, "message", "Error creating user: " + e.getMessage());
        }
    }
    
    public Map<String, Object> getAllUsers(String token) {
        try {
            // Check if current user is SUPERADMIN
            User currentUser = getCurrentUser(token);
            if (currentUser == null || currentUser.getRole() != UserRole.SUPERADMIN) {
                return Map.of("success", false, "message", "Access denied. Only SUPERADMIN can view all users.");
            }
            
            List<User> users = userRepository.findAll();
            List<Map<String, Object>> userList = new java.util.ArrayList<>();
            for (User user : users) {
                Map<String, Object> userMap = new HashMap<>();
                userMap.put("id", user.getId());
                userMap.put("username", user.getUsername());
                userMap.put("fullName", user.getFullName());
                userMap.put("role", user.getRole().name());
                userMap.put("department", user.getDepartment());
                userMap.put("isActive", user.getIsActive());
                userList.add(userMap);
            }
            
            return Map.of("success", true, "users", userList);
            
        } catch (Exception e) {
            System.out.println("DEBUG: Error getting users: " + e.getMessage());
            return Map.of("success", false, "message", "Error getting users: " + e.getMessage());
        }
    }
    
    public Map<String, Object> getAllUsersWithoutAuth() {
        try {
            System.out.println("DEBUG: Getting all users without authentication");
            List<User> users = userRepository.findAll();
            List<Map<String, Object>> userList = new java.util.ArrayList<>();
            for (User user : users) {
                Map<String, Object> userMap = new HashMap<>();
                userMap.put("id", user.getId());
                userMap.put("username", user.getUsername());
                userMap.put("fullName", user.getFullName());
                userMap.put("role", user.getRole().name());
                userMap.put("department", user.getDepartment());
                userMap.put("isActive", user.getIsActive());
                userList.add(userMap);
            }
            
            System.out.println("DEBUG: Found " + users.size() + " users");
            return Map.of("success", true, "users", userList);
            
        } catch (Exception e) {
            System.out.println("DEBUG: Error getting users without auth: " + e.getMessage());
            e.printStackTrace();
            return Map.of("success", false, "message", "Error getting users: " + e.getMessage());
        }
    }
    
    public Map<String, Object> updateUserStatus(String token, Long userId, Boolean isActive) {
        try {
            // Check if current user is SUPERADMIN
            User currentUser = getCurrentUser(token);
            if (currentUser == null || currentUser.getRole() != UserRole.SUPERADMIN) {
                return Map.of("success", false, "message", "Access denied. Only SUPERADMIN can update users.");
            }
            
            User user = userRepository.findById(userId).orElse(null);
            if (user == null) {
                return Map.of("success", false, "message", "User not found.");
            }
            
            user.setIsActive(isActive);
            userRepository.save(user);
            
            return Map.of("success", true, "message", "User status updated successfully");
            
        } catch (Exception e) {
            System.out.println("DEBUG: Error updating user status: " + e.getMessage());
            return Map.of("success", false, "message", "Error updating user status: " + e.getMessage());
        }
    }
}

File: frontend/src/app/services/auth.service.ts
import { Injectable, signal } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable, BehaviorSubject, of } from 'rxjs';
import { tap, catchError } from 'rxjs/operators';

export interface User {
  id: number;
  username: string;
  fullName: string;
  role: string;
  department: string;
}

export interface LoginResponse {
  success: boolean;
  token?: string;
  user?: User;
  message?: string;
}

@Injectable({
  providedIn: 'root'
})
export class AuthService {
  private baseUrl = '/api/auth';
  private currentUserSubject = new BehaviorSubject<User | null>(null);
  private tokenSubject = new BehaviorSubject<string | null>(null);

  public currentUser$ = this.currentUserSubject.asObservable();
  public token$ = this.tokenSubject.asObservable();

  constructor(private http: HttpClient) {
    // Check if user is already logged in (from localStorage)
    const savedToken = localStorage.getItem('token');
    const savedUser = localStorage.getItem('currentUser');
    
    if (savedToken && savedUser) {
      this.tokenSubject.next(savedToken);
      this.currentUserSubject.next(JSON.parse(savedUser));
    }
  }

  login(username: string, password: string): Observable<LoginResponse> 
  {
    const params = new HttpParams()
      .set('username', username)
      .set('password', password);

    console.log('DEBUG: Attempting login for user:', username);
    return this.http.post<LoginResponse>(`${this.baseUrl}/login`, null, { params })
      .pipe(
        tap(response => {
          console.log('DEBUG: Login response:', response);
          if (response.success && response.token && response.user) {
            this.tokenSubject.next(response.token);
            this.currentUserSubject.next(response.user);
            
            // Save to localStorage
            localStorage.setItem('token', response.token);
            localStorage.setItem('currentUser', JSON.stringify(response.user));
            console.log('DEBUG: Token saved:', response.token);
          } else {
            console.log('DEBUG: Login failed:', response.message);
          }
        })
      );
  }

  logout(): Observable<any> {
    const token = this.tokenSubject.value;
    
    // Always clear the token first
    this.clearSession();
    
    if (token) {
      const params = new HttpParams().set('token', token);
      return this.http.post(`${this.baseUrl}/logout`, null, { params })
        .pipe(
          tap(() => {
            console.log('Logout successful');
          }),
          catchError((error) => {
            console.error('Logout error:', error);
            // Even if backend logout fails, we've already cleared the token
            return of({ success: true });
          })
        );
    }
    return new Observable(observer => {
      observer.next({ success: true });
      observer.complete();
    });
  }

  getCurrentUser(): User | null {
    return this.currentUserSubject.value;
  }

  getToken(): string | null {
    const token = this.tokenSubject.value;
    console.log('DEBUG: getToken called, returning:', token);
    return token;
  }

  isLoggedIn(): boolean {
    return this.currentUserSubject.value !== null;
  }

  hasRole(role: string): boolean {
    const user = this.getCurrentUser();
    console.log('DEBUG hasRole: Checking role:', role);
    console.log('DEBUG hasRole: User role:', user?.role);
    console.log('DEBUG hasRole: Direct comparison:', user?.role === role);
    console.log('DEBUG hasRole: Is SUPERADMIN:', user?.role === 'SUPERADMIN');
    
    const result = user?.role === role || user?.role === 'SUPERADMIN';
    console.log('DEBUG hasRole: Final result:', result);
    return result;
  }

  canAccessITApprovals(): boolean {
    const result = this.hasRole('IT_MANAGER') || this.hasRole('SUPERADMIN');
    console.log('DEBUG canAccessITApprovals: hasRole(IT_MANAGER):', this.hasRole('IT_MANAGER'));
    console.log('DEBUG canAccessITApprovals: hasRole(SUPERADMIN):', this.hasRole('SUPERADMIN'));
    console.log('DEBUG canAccessITApprovals: result:', result);
    return result;
  }

  canAccessFinanceApprovals(): boolean {
    return this.hasRole('FINANCE_MANAGER') || this.hasRole('SUPERADMIN');
  }

  isDepartmentManager(): boolean {
    return this.hasRole('DEPARTMENT_MANAGER') || this.hasRole('SUPERADMIN');
  }

  canAccessBudget(): boolean {
    const user = this.getCurrentUser();
    console.log('DEBUG AuthService: Current user for budget check:', user);
    console.log('DEBUG AuthService: User role:', user?.role);
    console.log('DEBUG AuthService: Role type:', typeof user?.role);
    console.log('DEBUG AuthService: Is FINANCE_MANAGER?', user?.role === 'FINANCE_MANAGER');
    
    // Only FINANCE_MANAGER can access budget (SUPERADMIN excluded)
    const result = user?.role === 'FINANCE_MANAGER';
    console.log('DEBUG AuthService: Budget access result:', result);
    return result;
  }

  isEmployee(): boolean {
    const user = this.getCurrentUser();
    return user?.role === 'EMPLOYEE';
  }


  canCreateRequisitions(): boolean {
    // All logged-in users can create requisitions
    return this.isLoggedIn();
  }

  clearSession(): void {
    this.currentUserSubject.next(null);
    this.tokenSubject.next(null);
    localStorage.removeItem('token');
    localStorage.removeItem('currentUser');
  }
}
